DanmakuEngine.Tests/DearImgui/TestDrawDataBuffer.cs

```diff
@@ -76,4 +94,33 @@ public void TestPreTakeSnapShot_NotCollect()
 
         buffer.Dispose();
     }
+
+    [Test]
+    public void TestRealloc_EnlargeNotOverrideWhenCountLessThanAllocated()
+    {
+        // This test is to ensure that the old llocated lists are not overridden
+        // when the count is less than the allocated capacity
+        // This used to cause memory leaks so i added this test to ensure that it doesn't happen again
+
+        var buffer = new ImguiDrawDataBuffer(20, 0);
+
+        buffer.PreTakeSnapShot(15); // shrink count while not triggering gc
+
+        Debug.Assert(buffer.Count is 15);
+        Debug.Assert(buffer.Capacity is 20);
+
+        // Record addresses of the first 15 lists
+        IList<nint> _15Lists = new nint[15];
+
+        for (int i = 0; i < 15; i++)
+            _15Lists[i] = (nint)buffer.Lists[i];
+
+        buffer.PreTakeSnapShot(25); // enlarge count
+
+        Debug.Assert(buffer.Count is 25);
+
+        // Assert that our old lists are still there
+        for (int i = 0; i < 15; i++)
+            Assert.That((nint)buffer.Lists[i], Is.EqualTo(_15Lists[i]));
+    }
```

Comment:
The count and capacity are not what we should assert here, since we mainly focus on the previously allocated memory(if their addresses were overwritten).